From 1b173e40565889850f28b553101f86c8161d46ae Mon Sep 17 00:00:00 2001
From: Arulanbu Balusamy <quic_abalusam@quicinc.com>
Date: Mon, 24 Jun 2024 16:12:13 +0530
Subject: [PATCH] wifi: ath12k: Add the support to configure the legacy
 beacon tx rate.

Add the support to configure the beacon tx rate as legacy bit rate
This involves receiving the beacon rate mask from the mac80211 
layer and calculating the corresponding hardware rate code.

Patch-dependency: 920-002-wifi-nl80211-Used-the-chandef-from-the-configured-parameters-during.patch
		  703-01-cfg80211-add-support-for-set-bit-rate.patch
		  r34-hostapd-Fix-the-issue-while-configure-the-higher-leg.patch
		  r37-hostapd-Fix-NL80211_CMD_SET_BSS-failure-in-EHT-mode.patch

Patch-work: [None]

Signed-off-by: Arulanbu Balusamy <quic_abalusam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4457,14 +4457,17 @@ static void ath12k_recalculate_mgmt_rate
 {
 	const struct ieee80211_supported_band *sband;
 	u8 basic_rate_idx;
+	u8 beacon_rate_idx;
 	int hw_rate_code;
 	u32 vdev_param;
 	u16 bitrate;
 	int ret;
+	enum nl80211_band band;
 
 	lockdep_assert_held(&ar->conf_mutex);
 
 	sband = ar->ah->hw->wiphy->bands[def->chan->band];
+	band = def->chan->band;
 	basic_rate_idx = ffs(link_conf->basic_rates);
 	if (basic_rate_idx)
 		basic_rate_idx -= 1;
@@ -4482,6 +4485,18 @@ static void ath12k_recalculate_mgmt_rate
 	if (ret)
 		ath12k_warn(ar->ab, "failed to set mgmt tx rate %d\n", ret);
 
+	if (link_conf->beacon_tx_rate.control[band].legacy) {
+		beacon_rate_idx = ffs(link_conf->beacon_tx_rate.control[band].legacy);
+		beacon_rate_idx -=1;
+
+		if (band == NL80211_BAND_5GHZ || band == NL80211_BAND_6GHZ)
+			beacon_rate_idx += ATH12K_MAC_FIRST_OFDM_RATE_IDX;
+		if (beacon_rate_idx < ARRAY_SIZE(ath12k_legacy_rates)) {
+			bitrate = ath12k_legacy_rates[beacon_rate_idx].bitrate;
+			hw_rate_code = ath12k_mac_get_rate_hw_value(bitrate);
+		}
+	}
+
 	vdev_param = WMI_VDEV_PARAM_BEACON_RATE;
 	ret = ath12k_wmi_vdev_set_param_cmd(ar, arvif->vdev_id, vdev_param,
 					    hw_rate_code);
@@ -17296,6 +17311,8 @@ static int __ath12k_mac_register(struct
 
 	wiphy_ext_feature_set(hw->wiphy,
 			      NL80211_EXT_FEATURE_SET_SCAN_DWELL);
+	wiphy_ext_feature_set(hw->wiphy,
+			      NL80211_EXT_FEATURE_BEACON_RATE_LEGACY);
 
 	if (ar->pdev->cap.supported_bands & WMI_HOST_WLAN_5G_CAP) {
 		if (test_bit(ar->cfg_rx_chainmask, &cap->adfs_chain_mask)) {
